java-frontend/src/main/java/org/sonar/java/se/checks/NoWayOutLoopCheck.java

```diff
@@ -0,0 +1,229 @@
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.ForStatementTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.UnaryExpressionTree;
+import org.sonar.plugins.java.api.tree.WhileStatementTree;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+@Rule(
+  key = "S2189",
+  name = "Loops should not be infinite",
+  priority = Priority.BLOCKER,
+  tags = {"bug", "cert"})
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_RELIABILITY)
+@SqaleConstantRemediation("15min")
+public class NoWayOutLoopCheck extends SECheck {
+
+  private enum UpdateTypes {
+    INCREMENT, DECREMENT, INDETERMINATE
+  }
+
+  Map<Tree, CFGLoop> loopStarts;
+
+  @Override
+  public void init(MethodTree tree, CFG cfg) {
+    loopStarts = CFGLoop.getCFGLoops(cfg);
+  }
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    final PreStatementVisitor visitor = new PreStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  private class PreStatementVisitor extends CheckerTreeNodeVisitor {
+
+    private final CheckerContext context;
+
+    protected PreStatementVisitor(CheckerContext context) {
+      super(context.getState());
+      this.context = context;
+    }
+
+    @Override
+    public void visitWhileStatement(WhileStatementTree tree) {
+      if (isHardCodedTrue(tree.condition())) {
+        CFGLoop loopBlocks = loopStarts.get(tree);
+        if (loopBlocks.hasNoWayOut()) {
+          context.reportIssue(tree, NoWayOutLoopCheck.this, "Add an end condition to this loop.");
+        }
+      }
+    }
+
+    @Override
+    public void visitForStatement(ForStatementTree tree) {
+      CFGLoop loopBlocks = loopStarts.get(tree);
+      if (tree.condition() == null) {
+        if (loopBlocks.hasNoWayOut()) {
+          context.reportIssue(tree, NoWayOutLoopCheck.this, "Add an end condition to this loop.");
+        }
+      } else if (isConditionUnreachable(tree)) {
+        context.reportIssue(tree, NoWayOutLoopCheck.this, "Correct this loop's end condition.");
+      }
+    }
+
+    private boolean isConditionUnreachable(ForStatementTree tree) {
+      UpdatesCollector collector = new UpdatesCollector();
+      tree.accept(collector);
+      ConditionType condition = new ConditionType(tree.condition(), collector);
+      return !condition.isMatched();
+    }
+  }
+
+  private static class UpdateType {
```

Comment:
I would prefer to have this class called `Update` and the enum called `UpdateType` (without the plural). Having the enum and this class (which only does the binding between an update and a symbol) called by almost the same name is disturbing.